### PR TITLE
Support table-valued parameter

### PR DESCRIPTION
**Pull request by sphawk**
## What does this PR do?
Table-valued parameter implementation (for Microsoft SQL Server. aka MSSQL)

## What are related issues/pull requests?
detailed description on #255

## Environment
DB: MSSQL 2019
Driver: ODBC Driver 17 for SQL Server

### CODE REVIEW

**Comment by mloskot** [comment]
Is this ready for review?
If not, I think you can click the "Convert it to draft" to let us know it is not ready for review yet.

**Comment by sphawk** [comment]
Current progress:
- example\example_table_valued_parameter is working now. set -DNANODBC_ENABLE_TVP=ON, tested windows/linux and unicode enabled/disabled
- but I couldn't test on iODBC, because I don't know "Microsoft ODBC Driver 17 for SQL Server" working on iODBC, and I need wstring to u16string convert function for support iODBC. see https://docs.microsoft.com/en-us/sql/relational-databases/native-client-odbc-table-valued-parameters/binding-and-data-transfer-of-table-valued-parameters-and-column-values , BindParameter require tvp name as utf-16 string.

TODO:
- review class, variable names (nanodbc::tvp vs nanodbc::table_valued_parameter?)
- review implementation
- make more interfaces if we needs (like tvp::reset_parameters(), tvp::parameters(), ...)
- consider more features. like caching tvp name or parameters?
- make test
- make mssql test on travis-ci (linux to mssql), appveyor already has it.

may I convert to draft?

**Comment by mloskot** [comment]
I turned this PR into draft.

**Comment by mloskot** [comment]
> `nanodbc::tvp` vs `nanodbc::table_valued_parameter`

Please, avoid any abbreviations, prefer unambiguous long names.

> I can't imagine how many tests I should make

A minimum is to exercise every new functions added to the public interface.
Granularity is up to you, how you like to organise the test cases: you can have one, more or lots of test case functions, a test case per public function or a test case per feature or logical functionality.

**Comment by sphawk** [comment]
@mloskot  I'm ready for review.

File: nanodbc/nanodbc.cpp [hunk 1]
```diff
@@ -90,6 +90,37 @@
...
+#ifndef SQL_SS_NAME_SCOPE_DEFAULT
+#define SQL_SS_NAME_SCOPE_DEFAULT SQL_SS_NAME_SCOPE_TABLE
+#endif
+#ifndef SQL_CA_SS_BASE
+#define SQL_CA_SS_BASE 1200
+#endif
+#ifndef SQL_CA_SS_TYPE_NAME
+#define SQL_CA_SS_TYPE_NAME (SQL_CA_SS_BASE + 27)
+#endif
+
```
Comment (mloskot) [inline nanodbc/nanodbc.cpp:123]: A comment with source (i.e. is it  Windows SDK header file, what's the name) of these constants is necessary

**Review by mloskot — CHANGES_REQUESTED** [review]

File: test/mssql_test.cpp [hunk 2]
```diff
@@ -151,7 +215,7 @@ TEST_CASE_METHOD(mssql_fixture, "test_multi_statement_insert_select", "[mssql]")
 
     // Type of IDENTITY(seed,increment) return value is NUMERIC(38,0)
     // and the function may generate negative values too.
-    auto const sid = r.get<nanodbc::string>(0);
+    auto const sid = r.get<std::string>(0);
```
[the remark this answers is not in the document]
Comment (sphawk) [reply test/mssql_test.cpp:218]: no. but it's fix for compilation fail when NANODBC_ENABLE_UNICODE=ON
I found nanodbc::string is u16string on linux when NANODBC_ENABLE_UNICODE=ON
and std::stoll only take std::string and std::wstring

File: nanodbc/nanodbc.cpp (hunk 1 above)
Comment (sphawk) [reply nanodbc/nanodbc.cpp:125]: Add comments. I'll squash all commits, after review complete.

Comment (mloskot) [reply nanodbc/nanodbc.cpp:125]: No need to squash manually.
We hit the [Squash and commit](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges) button.

File: test/mssql_test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mloskot) [reply test/mssql_test.cpp:218]: We cover Unicode-based builds vs SQL Server on Windows

https://github.com/nanodbc/nanodbc/blob/cbc4ec2418a9feec559e06ea5dc0a632f305d5d1/.appveyor.yml#L28-L30

But, we don't seem to cover it on Linux

https://github.com/nanodbc/nanodbc/blob/cbc4ec2418a9feec559e06ea5dc0a632f305d5d1/.travis.yml

That may be why we haven't caught that issue.

**Review by mloskot — APPROVED** [review]
Although I'm happy to approve it, I'd prefer if we've had CI jobs in place that can exercise this feature against SQL Server on Linux. But, this will have to wait https://github.com/nanodbc/nanodbc/issues/265

**Comment by mloskot** [comment]
What's the status of this?
Would it be possible to update the PR description?

**Review by lexicalunit — APPROVED** [review]

**Comment by sphawk** [comment]
Added support for binding std::string_view in table_valued_parameter.bind_strings()
I'll make another PR if this accepted.

**Comment by mloskot** [comment]
Thank you

**Pull request merged**